File: 2386_find_the_k_sum_of_an_array/solution.go
// LeetCode 2386 - Find the K-Sum of an Array
// https://leetcode.com/problems/find-the-k-sum-of-an-array/

import (
	"container/heap"
	"sort"
)

type pair struct {
	sum int64
	i   int
}

type PQ []pair

func (h PQ) Len() int            { return len(h) }
func (h PQ) Less(i, j int) bool  { return h[i].sum > h[j].sum }
func (h PQ) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *PQ) Push(x interface{}) { *h = append(*h, x.(pair)) }
func (h *PQ) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

func kSum(nums []int, k int) int64 {
	var total int64
	absNums := make([]int, len(nums))
	for i, x := range nums {
		if x >= 0 {
			total += int64(x)
			absNums[i] = x
		} else {
			absNums[i] = -x
		}
	}
	sort.Ints(absNums)
	h := &PQ{{total, 0}}
	heap.Init(h)
	for t := 0; t < k-1; t++ {
		cur := heap.Pop(h).(pair)
		if cur.i >= len(absNums) {
			continue
		}
		heap.Push(h, pair{cur.sum - int64(absNums[cur.i]), cur.i + 1})
		if cur.i > 0 {
			heap.Push(h, pair{cur.sum - int64(absNums[cur.i]) + int64(absNums[cur.i-1]), cur.i + 1})
		}
	}
	return (*h)[0].sum
}

File: 2386_find_the_k_sum_of_an_array/solution.rb
# LeetCode 2386 - Find the K-Sum of an Array
# https://leetcode.com/problems/find-the-k-sum-of-an-array/

# @param {Integer[]} nums
# @param {Integer} k
# @return {Integer}
def k_sum(nums, k)
  total = 0
  n = nums.length
  abs_nums = Array.new(n, 0)
  (0...n).each do |i|
    if nums[i] >= 0
      total += nums[i]
      abs_nums[i] = nums[i]
    else
      abs_nums[i] = -nums[i]
    end
  end
  abs_nums.sort!
  h = []
  push = lambda do |item|
    h << item
    i = h.length - 1
    while i > 0
      p = (i - 1) >> 1
      break if h[p][0] >= h[i][0]
      h[p], h[i] = h[i], h[p]
      i = p
    end
  end
  pop = lambda do
    top = h[0]
    last = h.pop
    unless h.empty?
      h[0] = last
      i = 0
      loop do
        largest = i
        l = i * 2 + 1
        r = i * 2 + 2
        largest = l if l < h.length && h[l][0] > h[largest][0]
        largest = r if r < h.length && h[r][0] > h[largest][0]
        break if largest == i
        h[largest], h[i] = h[i], h[largest]
        i = largest
      end
    end
    top
  end
  push.call([total, 0])
  (k - 1).times do
    cur = pop.call
    s = cur[0]
    i = cur[1]
    next if i >= abs_nums.length
    push.call([s - abs_nums[i], i + 1])
    push.call([s - abs_nums[i] + abs_nums[i - 1], i + 1]) if i > 0
  end
  h[0][0]
end
